overlord/configstate/configcore/prompting.go

```diff
@@ -0,0 +1,89 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build !nomanagers
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"time"
+
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+)
+
+func doExperimentalApparmorPromptProfileRegeneration(c RunTransaction, opts *fsOnlyContext) error {
+	st := c.State()
+
+	snap, confName := features.AppArmorPrompting.ConfigOption()
+
+	var prompting bool
+	err := c.Get(snap, confName, &prompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	var prevPrompting bool
+	err = c.GetPristine(snap, confName, &prevPrompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if prompting == prevPrompting {
+		return nil
+	}
+
+	// XXX: what if apparmor-prompting flag value unchanged but support changed?
+	// AppArmor feature support checked once at startup, so this cannot occur.
+
+	if prompting && !features.AppArmorPrompting.IsSupported() {
+		// prompting newly enabled, but still not supported
+		return nil
+	}
+
+	st.Lock()
+	regenerateProfilesChg := st.NewChange("regenerate-all-security-profiles",
+		i18n.G("Regenerate all profiles due to change in prompting"))
+	snapSetupProfileTasks, err := ifacestate.CreateSnapSetupProfilesTasks(st)
+	if err != nil {
+		return err
+	}
+	usePromptPrefix := prompting && features.AppArmorPrompting.IsSupported()
+	for _, t := range snapSetupProfileTasks {
+		t.Set("use-prompt-prefix", &usePromptPrefix)
```

Comment:
Could we somehow inject this into snap setup profiles so that it lives in the same place as information about devmode or classic?

Reply:
The devmode and classic settings are in `confinementOptions` (are they elsewhere as well?). The PR adds a closure which we pass into `buildConfinementOptions`, and that (most of the time) checks to see if this value is set in the task. The reason to use the task is `buildConfinementOptions` might be called multiple times as part of the same change, and we want prompting to be consistently on or off for all confinement options related to that change.

But I agree, there's some redundancy here:
- If we don't define a custom closure, the default one will be created which:
  - first check if the task exists with the `"use-prompt-prefix"` value set, and if so, use that
  - otherwise, returns (supported && enabled), computed at time of closure creation (so it is consistent across invocations of the closure)
- If we do define a custom closure, we could have it ignore the task parameter and just return `true` or `false` based on some decision made at time of closure creation
  - However, we want to allow custom closures to have varying decisions based on the particular snap/task
  - This really is a niche use case though, I'm not sure when one would ever want to 

Basically, my current impulse is that optionally having both a task and/or a closure is redundant... My impulse is one would only ever have a task with the `"use-prompt-prefix"` value set if one has a custom closure, but then again, there could be a task created elsewhere when no closure is defined, and one *might* want to automatically cache the value on that task so the value is consistent with other `buildConfinementOptions` invocations with the same task.